crawlers/mooncrawl/mooncrawl/metadata_crawler/cli.py

```diff
@@ -78,39 +95,56 @@ def parse_metadata(blockchain_type: AvailableBlockchainType, batch_size: int):
     # run crawling of levels
     try:
 
-        uris_of_tokens = get_uris_of_tokens(db_session, blockchain_type)
-
-        tokens_uri_by_address: Dict[str, Any] = {}
+        meradata_addresses = get_uri_addresses(db_session, blockchain_type)
 
-        for token_uri_data in uris_of_tokens:
-            if token_uri_data.address not in tokens_uri_by_address:
-                tokens_uri_by_address[token_uri_data.address] = []
-            tokens_uri_by_address[token_uri_data.address].append(token_uri_data)
+        for address in meradata_addresses:
 
-        for address in tokens_uri_by_address:
+            not_updated_tokens = get_not_updated_metadata_for_address(
+                db_session,
+                blockchain_type,
+                address=address,
+            )
 
-            already_parsed = get_current_metadata_for_address(
-                db_session=db_session, blockchain_type=blockchain_type, address=address
+            logger.info(
+                f"Start crawling {len(not_updated_tokens)} tokens of address {address}"
             )
 
             for requests_chunk in [
-                tokens_uri_by_address[address][i : i + batch_size]
-                for i in range(0, len(tokens_uri_by_address[address]), batch_size)
+                not_updated_tokens[i : i + batch_size]
+                for i in range(0, len(not_updated_tokens), batch_size)
             ]:
                 writed_labels = 0
-                for token_uri_data in requests_chunk:
 
-                    if token_uri_data.token_id not in already_parsed:
-                        metadata = crawl_uri(token_uri_data.token_uri)
+                with ThreadPoolExecutor(
```

Comment:
With thread pool executor better to store results and `wait` them like in this code https://github.com/bugout-dev/moonstream/blob/e5240d2fd85685383bb40ea4cad741ec90b4fd6d/crawlers/mooncrawl/mooncrawl/blockchain.py#L367

Reply:
So this threads which inside another loop from list comprehension range will be under better control))